docs/source/introduction.md

```diff
@@ -531,6 +531,79 @@ You can also use a dictionary to specify either `dataset_options` or
 >>> write_dataframe(df, '/tmp/test.gpkg', dataset_options={"version": "1.0"}, layer_options={"geometry_name": "the_geom"})
 ```
 
+## Reading from and writing to in-memory datasets
+
+Pyogrio can automatically create and manage in-memory virtual files (`/vsimem`)
+to enable reading and writing from in-memory datasets.
+
+It is possible to read from a dataset stored as bytes:
+
+```python
+from io import BytesIO
+
+# save a GeoJSON to bytes
+geojson = """{
+    "type": "FeatureCollection",
+    "features": [
+        {
+            "type": "Feature",
+            "properties": { },
+            "geometry": { "type": "Point", "coordinates": [1, 1] }
+        }
+    ]
+}"""
+
+geojson_bytes = BytesIO(geojson.encode("UTF-8"))
+
+df = read_dataframe(geojson_bytes)
+```
+
+Note: this may emit a `RuntimeWarning` where the in-memory dataset is detected
+to be a particular format but lacks the expected file extension (e.g., `.gpkg`)
+because the in-memory path automatically created by pyogrio does not include the
+extension.
+
+It is also possible to write a dataset to bytes, but driver must also be
+specified, and layer name should be specified to avoid it being set to a random
+character string:
+
+```python
+buffer = BytesIO()
+
+write_dataframe(df, buffer, layer="my_layer", driver="GPKG")
+
+out_bytes = buffer.getvalue()
+```
+
+Note: this is limited to single-file data formats (e.g., GPKG) and does not
+support formats with multiple files (e.g., ESRI Shapefile).
+
+It is also possible to use a `/vsimem/` in-memory dataset with other GDAL-based
+packages that support the `/vsimem/` interface, such as the `gdal` package:
```

Comment:
I think it can also be useful to just use /vsimem without using other gdal-based packages... So would start with that (simpler) scenario... and then expand to combining with other gdal based packages?